src/components/Button/Button.tsx

```diff
@@ -1,155 +1,151 @@
+	inline?: boolean,
 	/**
 	 * Visual variations of the button.
 	 */
-	variant: PropTypes.oneOf( [
-		'primary',
-		'secondary',
-		'warning',
-	] ).isRequired,
+	variant?: Variant,
+} ) => {
+	const {
+		children,
+		inline = false,
+		variant = 'primary',
+	} = props;
+
+	return (
+		<span className={cx( 'label', `variant-${variant}`, { block: !inline } )}>{children}</span>
+	);
 };
 
 ButtonLabel.defaultProps = {
 	inline: false,
 	variant: 'primary',
 };
 
-const Button = ( {
-	ariaChecked,
-	children,
-	ariaDescribedBy,
-	disabled,
-	inline,
-	loading,
-	on,
-	onClick,
-	role,
-	type,
-	variant,
-} ) => (
-	<button
-		aria-checked={ariaChecked}
-		aria-describedby={ariaDescribedBy}
-		className={cx( 'button', { block: !inline } )}
-		disabled={disabled || loading}
-		on={on}
-		onClick={onClick}
-		role={role}
-		type={type}
-	>
-		<ButtonLabel variant={variant} inline={inline}>
-			{
-				loading && (
-					<div className={styles.spinner}>
-						<Spinner />
-					</div>
-				)
-			}
-			<span className={cx( 'label-text', { loading } )}>{children}</span>
-		</ButtonLabel>
-	</button>
-);
+type AriaRole = (
+	'checkbox' | 'link' | 'menuitem' | 'menuitemcheckbox' | 'menuitemradio' | 'option' | 'radio' | 'switch' | 'tab'
+)
 
-Button.propTypes = {
+const Button = ( props: {
 	/**
 	 * Forwarded to the button element as `aria-checked`. Only use when
 	 * `props.role` is one of `checkbox`, `menuitemcheckbox`,
 	 * `menuitemradiobutton`, `radiobutton`, or `switch`.
 	 */
-	ariaChecked: PropTypes.bool,
+	ariaChecked?: boolean,
 	/**
 	 * Forwarded to the button element as `aria-described-by`.
 	 */
-	ariaDescribedBy: PropTypes.string,
+	ariaDescribedBy?: string,
 	/**
 	 * Child nodes to be rendered as the label for the button. Maps to
 	 * the button element’s innerHTML.
 	 */
-	children: PropTypes.node.isRequired,
+	children: React.ReactNode,
 	/**
 	 * Whether the button is interactive. Forwarded to the button
 	 * element.
 	 */
-	disabled: PropTypes.bool.isRequired,
+	disabled?: boolean,
+	/**
+	 * Same visual state as `props.disabled`, but without affecting
+	 * interactivity. Use this when you want to indicate that a form's
+	 * contents are invalid, but still submissible.
+	 */
+	inactive?: boolean,
 	/**
 	 * Whether the button should be purely textual, e.g. for use in a
 	 * paragraph of text.
 	 */
-	inline: PropTypes.bool.isRequired,
+	inline?: boolean,
 	/**
 	 * Visually replaces `props.children` with the Spinner component. Use
 	 * when waiting for an action to complete in response to a user
 	 * engaging the button.
 	 */
-	loading: PropTypes.bool.isRequired,
+	loading?: boolean,
 	/**
 	 * For binding Google AMP interactivity. See
 	 * https://amp.dev/documentation/examples/components/amp-bind/
 	 */
-	on: PropTypes.string,
+	on?: string,
 	/**
 	 * Function to run when the button is engaged. Forwarded to the button
 	 * element.
 	 */
-	onClick: PropTypes.func.isRequired,
+	onClick?: React.UIEventHandler<HTMLButtonElement>,
```

Comment:
Hmm - why make onClick optional? It seems like a useful pointer to say that maybe another kind of element is required

Reply:
A `<button type="submit">` doesn't need (and shouldn't have) an `onClick`.

I think this nit should be obvious when _calling_ the component. But when I converted to Typescript, I erroneously set this to "required" and then had to backtrack when the callers threw type-safety errors. So I've added an inline comment to save the next developer the same embarrassment.

Reply:
🤦 of course! Good points.